### PR TITLE
feat - new promise based api

### PR DESCRIPTION
**Pull request by StarpTech**
This is a preview of the new interface. I appreciate any feedback. Tests suite is incomplete but the features are working already.

Checkout the `examples` to get a better impression or play with the `form.html`

Related: #109, #139 

#### Checklist

- [X] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by jsumners** [comment]
I rather prefer the original interface. This strikes me as too inflexible.

**Comment by StarpTech** [comment]
Could you be more specific? What do you miss?

File: test/big.test.js [hunk 1]
```diff
@@ -1,112 +0,0 @@
...
-const http = require('http')
-const stream = require('readable-stream')
-const Readable = stream.Readable
-const Writable = stream.Writable
-const pump = stream.pipeline
-const eos = stream.finished
-const crypto = require('crypto')
-
-// skipping on Github Actions because it takes too long
-test('should upload a big file in constant memory', { skip: process.env.CI }, function (t) {
```
Comment (mcollina) [inline test/big.test.js:16]: This test is actually important to check if we do not cause a memory leak.

File: index.js [hunk 2]
```diff
@@ -87,122 +32,213 @@ function busboy (options) {
...
 
-    const busboyOptions = deepmerge.all([{ headers: req.headers }, options || {}, opts || {}])
-    const stream = busboy(busboyOptions)
-    var completed = false
-    var files = 0
-    var count = 0
-    var callDoneOnNextEos = false
+    function onField (name, fieldValue, fieldnameTruncated, valueTruncated) {
+      // don't overwrite prototypes
+      if (getDescriptor(Object.prototype, name)) return
```
Comment (mcollina) [inline index.js:122]: Should we error for this instead?

File: index.js [hunk 3]
```diff
@@ -87,122 +32,213 @@ function busboy (options) {
...
+      request.removeListener('close', cleanup)
+      bb.removeListener('field', onField)
+      bb.removeListener('file', onFile)
+      bb.removeListener('close', cleanup)
+    }
+
+    return parts
+  }
+
+  async function saveRequestFiles () {
```
Comment (mcollina) [inline index.js:186]: I think this should throw if some of the other methods are called first.

File: test/append-body.test.js [hunk 4]
```diff
@@ -1,788 +0,0 @@
...
-const Fastify = require('fastify')
-const multipart = require('..')
-const http = require('http')
-const path = require('path')
-const fs = require('fs')
-const pump = require('pump')
-
-const filePath = path.join(__dirname, '../README.md')
-
-test('addToBody option', t => {
```
Comment (mcollina) [inline test/append-body.test.js:13]: the `addToBody` option is needed to support automatic validation of the body itself if a user would like to do so. If we remove this functionality, we should add some equivalent replacement.

**Comment by jsumners** [comment]
> Could you be more specific? What do you miss?

It looks to me like this implementation assumes "file" uploads and always writes the data out to a file on the filesystem. Multipart isn't always about uploading files. In particular, I have a case where I need to look for a part named "json" and read the data attached to it as an object; I have no need of writing this to the filesystem.

File: test/big.test.js (hunk 1 above)
Comment (StarpTech) [reply test/big.test.js:16]: Don't worry I just remove it for a clean start. I will add this when we will agree on the implementation.

File: index.js (hunk 2 above)
Comment (StarpTech) [reply index.js:122]: I agree.

**Comment by StarpTech** [comment]
> It looks to me like this implementation assumes "file" uploads and always writes the data out to a file on the filesystem. Multipart isn't always about uploading files. In particular, I have a case where I need to look for a part named "json" and read the data attached to it as an object; I have no need of writing this to the filesystem.

Oh no, please check out all examples. You can work with just streams or files.

File: index.js (hunk 3 above)
Comment (StarpTech) [reply index.js:186]: Good point.

File: index.js [hunk 5]
```diff
@@ -87,122 +32,213 @@ function busboy (options) {
...
 
+  function handleMultipart (opts = {}) {
     if (!this.isMultipart()) {
-      done(new Error('the request is not multipart'))
-      return
+      throw new Error('the request is not multipart')
     }
 
-    const log = this.log
+    const ch = chan()
```
[the remark this answers is not in the document]
Comment (StarpTech) [reply index.js]: I replaced this module with a simple FIFO. Please check.

**Comment by jsumners** [comment]
> It also looks a lot like a new module to me.

At a minimum, I agree with this statement. One reason I think a major framework has recently been deprecated is due to the way it was routinely using semver major to release what amounted to new frameworks under the same name. This PR completely changes the API for this module, and will be a significant hinderance to people using the module. I think that is a worse developer experience than the current API.

**Comment by StarpTech** [comment]
Wow, why so negative? Let focus at first on the PR. We can discuss afterward how we can incorporate that in an fastify ecosystem.

**Comment by jsumners** [comment]
I'm not seeing any negativity in my statement. I only stated that I agree _this_ module should remain as-is.

**Comment by StarpTech** [comment]
> I'm not seeing any negativity in my statement. I only stated that I agree this module should remain as-is.

You made it already clear with your first [comment](https://github.com/fastify/fastify-multipart/pull/140#issuecomment-671115459).

File: index.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (StarpTech) [reply index.js]: Check again I could simplify it.

File: index.js [hunk 6]
```diff
@@ -87,122 +33,235 @@ function busboy (options) {
...
-    fastify.addHook('preValidation', function (req, reply, next) {
-      attachToBody(options, req, reply, next)
-    })
-  }
+function fastifyMultipart (fastify, options = {}, done) {
   fastify.addContentTypeParser('multipart', setMultipart)
-
-  fastify.decorateRequest('multipart', multipart)
+  fastify.decorateRequest('handleMultipart', handleMultipart)
+  fastify.decorateRequest('multipart', getMultipartIterator)
```
Comment (Eomm) [inline index.js:39]: If we leave this prop with as the old interface `const bb = req.multipart(handler, onEnd)` could be a solution @jsumners ?

Comment (jsumners) [reply index.js:39]: I think so. But it might be difficult to support both interfaces.

Comment (StarpTech) [reply index.js:39]: The new interface provides the same flexibility: Consuming streams. On top of that, we can work with files directly and support async/await.

IMO it's not worth supporting both interfaces or what do you see that I don't see?

**Comment by StarpTech** [comment]
@mcollina on windows the tests `npm test` will run in a timeout. When I run the tests via `node` they will fail only occasionally.

E.g it will be stuck.

```
$ node test/multipart-stream.test.js 
TAP version 13
# Subtest: should throw fileSize limitation error on small payload
    1..4
    ok 1 - expect truthy value
    ok 2 - should be equal

```

It looks like it's not related to a specific test. On my machine `'should throw fileSize limitation error on small payload'` and `'should not allow __proto__ as file name'` fail regularly. Perhaps you find the bug in the code.

**Comment by StarpTech** [comment]
@mcollina I could fix it 😄

**Comment by mcollina** [comment]
go for a fix! I don't have a clue - the first is likely due to difference in the network stack.

**Comment by StarpTech** [comment]
On windows ci `npm install` does not even start 🤯 On MAC the tests failing. On Linux works everything fine. I need some help.

**Comment by mcollina** [comment]
> On windows ci npm install does not even start

I do not have any idea about that.

---

I'll take a look for Mac.

**Comment by StarpTech** [comment]
This was super confusing. The other ci were canceled due to `fail-fast` matrix strategy. Now I have a `UnhandledPromiseRejectionWarning: Error: write EPIPE` on Windows and MAC but I can't reproduce it locally.

**Comment by StarpTech** [comment]
@mcollina It looks like that the server closes the request on a file stream error but I have no clue why.

**Comment by StarpTech** [comment]
I found the issue. The stream was not drained after I called `reply.send`.

**Comment by StarpTech** [comment]
I'm on vacation next week. Feel free to review the current API. After that, I will try to reintegrate the old one.

**Review by fox1t — COMMENTED** [review]
I really like the new API. 👍 
I think that makes this plugin more usable without losing any of its strengths.

File: test/append-body.test.js (hunk 4 above)
Comment (StarpTech) [reply test/append-body.test.js:13]: JSON Schema was added.

File: index.js [hunk 7]
```diff
@@ -1,11 +1,22 @@
-'use strict'
```
Comment (mcollina) [inline index.js:1]: use strict is still needed

File: index.js [hunk 8]
```diff
@@ -200,9 +275,250 @@ function fastifyMultipart (fastify, options, done) {
...
+        buffer () {
+          if (this._buf) {
+            return Promise.resolve(this._buf)
+          }
+          return new Promise((resolve, reject) => {
+            pump(this.file, concat((buf) => {
+              this._buf = buf
+              resolve(this._buf)
+            })).catch(reject)
+          })
```
Comment (mcollina) [inline index.js:395]: I would avoid breaking the promise chain for this. Maybe we can use `this.file` async iterator instead?

File: index.js [hunk 9]
```diff
@@ -200,9 +275,250 @@ function fastifyMultipart (fastify, options, done) {
...
+    file.once('limit', () => {
+      const err = new RequestFileTooLargeError()
+
+      if (file.listenerCount('error') > 0) {
+        file.emit('error', err)
+        logger.warn(err)
+      } else {
+        logger.error(err)
+        // ignore next error event
+        file.on('error', () => { })
```
Comment (mcollina) [inline index.js:450]: why is this ignored? Maybe we should log it as debug?

File: index.js [hunk 10]
```diff
@@ -200,9 +230,261 @@ function fastifyMultipart (fastify, options, done) {
...
+
+    bb
+      .on('field', onField)
+      .on('file', onFile)
+      .on('close', cleanup)
+      .on('error', onEnd)
+      .on('finish', onEnd)
+
+    bb.on('partsLimit', function () {
+      const err = new Error('Reach parts limit')
```
[the remark this answers is not in the document]
Comment (StarpTech) [reply index.js]: You refer to an old commit. This is already done.

File: index.js [hunk 11]
```diff
@@ -200,9 +230,261 @@ function fastifyMultipart (fastify, options, done) {
...
+      err.code = 'Request_parts_limit'
+      err.status = 413
+      onError(err)
+    })
+
+    bb.on('filesLimit', function () {
+      const err = new Error('Reach files limit')
+      err.code = 'Request_files_limit'
```
[the remark this answers is not in the document]
Comment (StarpTech) [reply index.js]: Same

File: index.js (hunk 9 above)
Comment (StarpTech) [reply index.js:450]: I will log it as an error.

File: index.js (hunk 8 above)
Comment (StarpTech) [reply index.js:395]: Good point.

**Review by mcollina — APPROVED** [review]
lgtm, let's wait for the others anyway.

**Comment by StarpTech** [comment]
> lgtm, let's wait for the others anyway.

@mcollina I will wait for one more. The activity here is extremely low. I'd like to use this module.

File: README-legacy.md [hunk 12]
```diff
@@ -0,0 +1,194 @@
+# Callback-based API
```
Comment (mcollina) [inline README-legacy.md:1]: Can you rename this file to `callback.md`?

Comment (StarpTech) [reply README-legacy.md]: Done.

**Comment by mcollina** [comment]
Can you change the title of this PR?

**Review by delvedor — APPROVED** [review]
LGTM

**Pull request merged**